// @airtable/blocks/dist/esm/base/index.js
import './assert_run_context';
import { __injectSdkIntoWarning } from '../shared/warning';
import getAirtableInterface from '../injected/airtable_interface';
import { __injectSdkIntoPerformRecordAction } from './perform_record_action';
import Sdk from './sdk';
import { __injectSdkIntoInitializeBlock } from './ui/initialize_block';

/** @internal */
export let __sdk;
export let base;
export let globalConfig;
export let installationId;
export let reload;
export let runInfo;
export let settingsButton;
export let undoRedo;
export let viewport;
export let unstable_fetchAsync;

/** @internal */
export function __reset() {
  __sdk = new Sdk(getAirtableInterface());
  ({
    base,
    globalConfig,
    installationId,
    reload,
    runInfo,
    settingsButton,
    undoRedo,
    viewport,
    unstable_fetchAsync
  } = __sdk);

  // The following injections are necessary to allow the referenced modules
  // to be imported directly by consumer code while also avoiding cycles in
  // the module dependency graph.
  __injectSdkIntoPerformRecordAction(__sdk);
  __injectSdkIntoInitializeBlock(__sdk);
  __injectSdkIntoWarning(__sdk);
}
__reset();
