// @airtable/blocks/dist/esm/base/types/mutations.js
/** @module @airtable/blocks: mutations */ /** */

import { MutationTypesCore } from '../../shared/types/mutations_core';
/** @hidden */
export const MutationTypes = Object.freeze({
  ...MutationTypesCore,
  CREATE_SINGLE_FIELD: 'createSingleField',
  UPDATE_SINGLE_FIELD_CONFIG: 'updateSingleFieldConfig',
  UPDATE_SINGLE_FIELD_DESCRIPTION: 'updateSingleFieldDescription',
  UPDATE_SINGLE_FIELD_NAME: 'updateSingleFieldName',
  CREATE_SINGLE_TABLE: 'createSingleTable',
  UPDATE_VIEW_METADATA: 'updateViewMetadata'
});

/** @hidden */

/**
 * The Mutation emitted when the App creates a {@link Field}.
 *
 * @docsPath testing/mutations/CreateSingleFieldMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies the options of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldConfigMutation
 */
// note: while this accepts both type and options in config, we enforce liveapp-side that type
// is not changed.

/** @hidden */

/**
 * The Mutation emitted when the App modifies the description of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldDescriptionMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies the name of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldNameMutation
 */

/** @hidden */

/**
 * Options that affect the behavior of an `UpdateSingleFieldConfigMutation`
 */

/**
 * The Mutation emitted when the App creates a {@link Table}
 *
 * @docsPath testing/mutations/CreateSingleTableMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies a {@link View}.
 *
 * @docsPath testing/mutations/UpdateViewMetadataMutation
 */

/** @hidden */

/** @hidden */

/** @hidden */
