// @airtable/blocks/dist/esm/stats/block_stats.js
import getAirtableInterface from '../injected/airtable_interface';

// NOTE(richsinn): this client is effectively an implementation of the StatsClient interface
// for Datadog. For example, see https://github.com/Hyperbase/hyperbase/blob/development/client_server_shared/stats/stats_client.tsx

/** @hidden */
class BlockStats {
  increment(stat, tags) {
    this.incrementBy(stat, 1, tags);
  }
  incrementBy(stat, value, tags) {
    getAirtableInterface().sendStat({
      metricType: 'incrementBy',
      stat,
      value,
      tags
    });
  }
  decrement(stat, tags) {
    this.decrementBy(stat, 1, tags);
  }
  decrementBy(stat, value, tags) {
    getAirtableInterface().sendStat({
      metricType: 'decrementBy',
      stat,
      value,
      tags
    });
  }
  gauge(stat, value, tags) {
    getAirtableInterface().sendStat({
      metricType: 'gauge',
      stat,
      value,
      tags
    });
  }
  histogram(stat, value, tags) {
    getAirtableInterface().sendStat({
      metricType: 'histogram',
      stat,
      value,
      tags
    });
  }
  distribution(stat, value, tags) {
    getAirtableInterface().sendStat({
      metricType: 'distribution',
      stat,
      value,
      tags
    });
  }
  timing(stat, time, tags) {
    getAirtableInterface().sendStat({
      metricType: 'timing',
      stat,
      value: time,
      tags
    });
  }
  timingWithSum(stat, time, tags) {
    this.timing(stat, time, tags);
    this.incrementBy(stat + '.sum', time, tags);
  }
}
export const blockStats = new BlockStats();
