// @airtable/blocks/dist/esm/shared/ui/cell_renderer.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
/** @module @airtable/blocks/ui: CellRenderer */ /** */
import * as React from 'react';
import { spawnError } from '../../shared/error_utils';
import { FieldType } from '../../shared/types/field_core';
import useWatchable from '../../shared/ui/use_watchable';
import { useSdk } from '../../shared/ui/sdk_context';

/**
 * @hidden
 */

/**
 * @internal
 */
function validateRecordAndFieldProps(props) {
  if (props.record && !props.record.isDeleted && !props.field.isDeleted && props.record.parentTable.id !== props.field.parentTable.id) {
    throw spawnError('CellRenderer: record %s and field %s do not have the same parent table', props.record.parentTable.id, props.field.parentTable.id);
  }
}

/**
 * @hidden
 */
export function CellRenderer(props) {
  const {
    record,
    cellValue,
    field,
    shouldWrap = true,
    onMouseEnter,
    onMouseLeave,
    onClick,
    className,
    style,
    cellClassName,
    cellStyle,
    renderInvalidCellValue
  } = props;
  validateRecordAndFieldProps({
    record,
    field
  });
  const sdk = useSdk();
  useWatchable(record, [`cellValueInField:${field.id}`]);
  useWatchable(field, ['type', 'options']);
  if (field.isDeleted) {
    return null;
  }
  const airtableInterface = sdk.__airtableInterface;
  const appInterface = sdk.__appInterface;
  let cellValueToRender;
  if (record) {
    if (cellValue !== undefined) {
      // eslint-disable-next-line
      console.warn('CellRenderer was given both record and cellValue, choosing to render record value');
    }
    if (record.isDeleted) {
      return null;
    }
    cellValueToRender = record.getCellValue(field.id);
  } else {
    // NOTE: this will not work if you want to render a cell value for
    // foreign record, single/multi select, or single/multi collaborator
    // fields and the cell value is not *currently* valid for that field.
    // i.e. if you want to render a foreign record for a record that
    // does not yet exist, this will throw.
    // TODO: handle "preview" cell values that are not yet valid in the given field
    // but that *could* be.
    // To validate public cell values, we only have validateCellValueForUpdate
    // However, this is not implemented for computed fields (since you can't update them)
    // so we just skip the check.
    // TODO(emma): actually check this somehow.
    if (!field.isComputed) {
      const validationResult = airtableInterface.fieldTypeProvider.validateCellValueForUpdate(appInterface, cellValue, null, field._data);
      if (!validationResult.isValid) {
        if (renderInvalidCellValue) {
          return /*#__PURE__*/React.createElement("div", {
            onMouseEnter: onMouseEnter,
            onMouseLeave: onMouseLeave,
            onClick: onClick,
            className: className,
            style: style
          }, renderInvalidCellValue(cellValue, field));
        } else {
          throw spawnError('Cannot render invalid cell value %s: %s', cellValue, validationResult.reason);
        }
      }
    }
    cellValueToRender = cellValue;
  }

  // in record.getCellValue, we reformat the cell value to the new public format. in order to
  // make it work with the cell renderer, we need to format it back the other way.
  // Note: we can't use _getRawCellValue here since we also need to handle formatting
  // these in the case where the user passed cellValue instead of a record.
  if (cellValueToRender && field.type === FieldType.MULTIPLE_LOOKUP_VALUES && !airtableInterface.sdkInitData.isUsingNewLookupCellValueFormat) {
    const originalCellValue = cellValueToRender;
    const linkedRecordIdsSet = new Set();
    const valuesByLinkedRecordId = {};
    for (const {
      linkedRecordId,
      value
    } of originalCellValue) {
      linkedRecordIdsSet.add(linkedRecordId);
      if (!valuesByLinkedRecordId[linkedRecordId]) {
        valuesByLinkedRecordId[linkedRecordId] = [];
      }
      valuesByLinkedRecordId[linkedRecordId].push(value);
    }
    cellValueToRender = {
      linkedRecordIds: Array.from(linkedRecordIdsSet),
      valuesByLinkedRecordId
    };
  }
  const {
    cellValueHtml,
    attributes
  } = airtableInterface.fieldTypeProvider.getCellRendererData(appInterface, cellValueToRender, field._data, !!shouldWrap);
  return /*#__PURE__*/React.createElement("div", {
    // TODO (stephen): remove tooltip anchor props
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    className: `baymax ${className ?? ''}`,
    style: {
      display: 'block',
      ...style
    }
  }, /*#__PURE__*/React.createElement("div", _extends({}, attributes, {
    className: `cell read ${cellClassName ?? ''}`,
    style: cellStyle,
    dangerouslySetInnerHTML: {
      __html: cellValueHtml
    }
  })));
}
