// @airtable/blocks/dist/esm/base/ui/block_wrapper.js
/** @hidden */ /** */
import * as React from 'react';
import { SdkContext } from '../../shared/ui/sdk_context';
import Loader from '../../shared/ui/loader';
import { globalAlert, useWatchable } from './ui';
import { getCssContentToAddToHead, BUTTON_LINK_CLASS_NAME, SPIN_SCALE_ANIMATION_NAME } from './global_css_helpers';
const suspenseFallbackStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
};
const wrapperStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  overflow: 'hidden'
};
const fullscreenMessageStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '1rem',
  backgroundColor: '#ffffff',
  zIndex: 2147483647 // largest 32-bit signed integer (maximum z-index)
};
const animateSpinnerStyle = {
  animationIterationCount: 'infinite',
  animationName: SPIN_SCALE_ANIMATION_NAME,
  animationDuration: '1800ms',
  animationTimingFunction: 'cubic-bezier(0.785, 0.135, 0.15, 0.86)'
};
export default function BlockWrapper(props) {
  const {
    sdk,
    children
  } = props;
  const [globalAlertInfoIfExists, setGlobalAlertInfoIfExists] = React.useState(null);
  React.useEffect(() => {
    // Once we have a global alert info, we don't need to watch for changes. The
    // global alert will display until the page is reloaded.
    if (globalAlertInfoIfExists) {
      return () => {};
    }

    // Although we call globalAlert.watch('__alertInfo', ..) below, this will only
    // start watching after the component is mounted. If some child component called
    // globalAlert.showReloadPrompt in its constructor or useEffect (recall that child
    // useEffects run before parent useEffects), then __alertInfo would be set before
    // we start watching it here. Therefore explicitly check if __alertInfo is set.
    if (globalAlert.__alertInfo) {
      setGlobalAlertInfoIfExists(globalAlert.__alertInfo);
    }

    // Watch for changes to the global alert info.
    const onAlertInfoChange = a => setGlobalAlertInfoIfExists(a.__alertInfo);
    globalAlert.watch('__alertInfo', onAlertInfoChange);
    return () => {
      globalAlert.unwatch('__alertInfo', onAlertInfoChange);
    };
  }, [globalAlertInfoIfExists]);
  const viewport = sdk.viewport;
  useWatchable(viewport, ['size', 'minSize']);
  // Although we watch for changes to minSize above, if a child component
  // sets minSize in its constructor or useEffect, it can change before useWatchable
  // has a chance to set up watches.
  // To get around this, we take a snapshot of the minSize before render and call
  // .forceUpdate() if its changed after render.
  const minSizeBeforeFirstRender = React.useRef(viewport.minSize);
  const [, forceUpdate] = React.useReducer(x => x + 1, 0);
  React.useEffect(() => {
    if (minSizeBeforeFirstRender.current) {
      const prevMinSize = minSizeBeforeFirstRender.current;
      minSizeBeforeFirstRender.current = null;
      const currentMinSize = viewport.minSize;
      if (currentMinSize.width !== prevMinSize.width || currentMinSize.height !== prevMinSize.height) {
        forceUpdate();
      }
    }
  }, [viewport]);
  React.useLayoutEffect(() => {
    // Create a style element and inject the global CSS into document head
    const styleElement = document.createElement('style');
    styleElement.textContent = getCssContentToAddToHead();
    document.head.appendChild(styleElement);

    // Cleanup: remove the style element when component unmounts
    return () => {
      if (document.head.contains(styleElement)) {
        document.head.removeChild(styleElement);
      }
    };
  }, []);
  if (globalAlertInfoIfExists) {
    return /*#__PURE__*/React.createElement(SdkContext.Provider, {
      value: sdk
    }, /*#__PURE__*/React.createElement("div", {
      style: fullscreenMessageStyle
    }, globalAlertInfoIfExists.content));
  }
  return /*#__PURE__*/React.createElement(SdkContext.Provider, {
    value: sdk
  }, /*#__PURE__*/React.createElement(React.Suspense, {
    fallback: /*#__PURE__*/React.createElement("div", {
      style: suspenseFallbackStyle
    }, /*#__PURE__*/React.createElement(Loader, {
      style: animateSpinnerStyle
    }))
  }, /*#__PURE__*/React.createElement("div", {
    style: viewport.isSmallerThanMinSize ? wrapperStyle : undefined
  }, children)), viewport.isSmallerThanMinSize && /*#__PURE__*/React.createElement(MakeBiggerOrFullscreenMessage, {
    viewport: viewport
  }));
}
function MakeBiggerOrFullscreenMessage(_ref) {
  let {
    viewport
  } = _ref;
  return /*#__PURE__*/React.createElement("div", {
    style: fullscreenMessageStyle
  }, /*#__PURE__*/React.createElement("span", {
    style: {
      textAlign: 'center',
      lineHeight: 1.5,
      opacity: 0.75,
      fontWeight: 500
    }
  }, /*#__PURE__*/React.createElement("span", null, "Please make this extension bigger or "), /*#__PURE__*/React.createElement("span", {
    className: BUTTON_LINK_CLASS_NAME,
    onClick: () => viewport.enterFullscreenIfPossible()
  }, "fullscreen")));
}
