// @airtable/blocks/dist/esm/base/ui/global_alert.js
/** @hidden */ /** */
import * as React from 'react';
import { isEnumValue } from '../../shared/private_utils';
import Watchable from '../../shared/watchable';
import { useSdk } from '../../shared/ui/sdk_context';
import { BUTTON_LINK_CLASS_NAME } from './global_css_helpers';
const WatchableGlobalAlertKeys = Object.freeze({
  __alertInfo: '__alertInfo'
});
/** @hidden */

/** @hidden */

const GlobalAlertInfo = () => {
  const sdk = useSdk();
  return /*#__PURE__*/React.createElement("span", {
    style: {
      textAlign: 'center',
      lineHeight: 1.5,
      fontWeight: 500
    }
  }, /*#__PURE__*/React.createElement("span", {
    className: BUTTON_LINK_CLASS_NAME,
    onClick: () => {
      sdk.reload();
    }
  }, "Please reload"));
};

/**
 * @hidden
 * @example
 * ```js
 * import {globalAlert} from '@airtable/blocks/base/ui';
 * globalAlert.showReloadPrompt();
 * ```
 */
class GlobalAlert extends Watchable {
  /** @internal */
  static _className = 'GlobalAlert';
  /** @internal */
  static _isWatchableKey(key) {
    return isEnumValue(WatchableGlobalAlertKeys, key);
  }
  /** @internal */

  /** @hidden */
  constructor() {
    super();
    this._alertInfo = null;
    Object.seal(this);
  }
  /** @internal */
  get __alertInfo() {
    return this._alertInfo;
  }
  /** */
  showReloadPrompt() {
    this._alertInfo = {
      content: /*#__PURE__*/React.createElement(GlobalAlertInfo, null)
    };
    this._onChange(WatchableGlobalAlertKeys.__alertInfo);
  }
}
export default GlobalAlert;
