// @airtable/blocks/dist/esm/base/ui/initialize_block.js
/** @module @airtable/blocks/ui: initializeBlock */ /** */
import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { spawnError } from '../../shared/error_utils';
import getAirtableInterface from '../../injected/airtable_interface';
import { BlockRunContextType } from '../types/airtable_interface';
import BlockWrapper from './block_wrapper';
let hasBeenInitialized = false;

/** */

/** @hidden */

/** @hidden */

/** @hidden */

/**
 * `initializeBlock` takes the top-level React component in your tree and renders it. It is conceptually similar to `ReactDOM.render`, but takes care of some Extensions-specific things.
 *
 * @param getEntryElement A function that returns your React Node.
 *
 * @example
 * ```js
 * import {initializeBlock} from '@airtable/blocks/base/ui';
 * import React from 'react';
 *
 * function App() {
 *     return (
 *         <div>Hello world 🚀</div>
 *     );
 * }
 *
 * initializeBlock(() => <App />);
 * ```
 * @docsPath UI/utils/initializeBlock
 */
export function initializeBlock(getEntryElement) {
  const entryPoints = typeof getEntryElement === 'function' ? {
    dashboard: getEntryElement
  } : getEntryElement;
  const body = typeof document !== 'undefined' ? document.body : null;
  if (!body) {
    throw spawnError('initializeBlock should only be called from browser environments');
  }
  if (hasBeenInitialized) {
    throw spawnError('initializeBlock should only be called once');
  }
  hasBeenInitialized = true;
  const airtableInterface = getAirtableInterface();
  let entryElement;
  // runContext can be undefined if running from an old version client version (before 01/2021)
  // TODO (SeanKeenan): Remove nullish coelescing once old clients are no longer a concern
  const runContext = airtableInterface.sdkInitData.runContext ?? {
    type: BlockRunContextType.DASHBOARD_APP
  };
  switch (runContext.type) {
    case BlockRunContextType.DASHBOARD_APP:
      {
        if (entryPoints.dashboard === undefined) {
          throw spawnError('If running an extension within the dashboard, it must have a dashboard initialization function');
        }
        if (typeof entryPoints.dashboard !== 'function') {
          throw spawnError('initializeBlock must contain a dashboard function that returns a React element');
        }
        entryElement = entryPoints.dashboard();
        break;
      }
    case BlockRunContextType.VIEW:
      {
        if (entryPoints.view === undefined) {
          throw spawnError('If running an extension within a view, it must have a view initialization function');
        }
        if (typeof entryPoints.view !== 'function') {
          throw spawnError('initializeBlock must contain a view function that returns a React element');
        }
        const table = sdk.base.getTableById(runContext.tableId);
        const view = table.getViewById(runContext.viewId);
        entryElement = entryPoints.view({
          table,
          view
        });
        break;
      }
    default:
      throw spawnError('Invalid context to run ');
  }
  if (! /*#__PURE__*/React.isValidElement(entryElement)) {
    throw spawnError("The first argument to initializeBlock didn't return a valid React element");
  }
  const container = document.createElement('div');
  container.style.height = '100%';
  container.style.width = '100%';
  body.appendChild(container);
  createRoot(container).render( /*#__PURE__*/React.createElement(BlockWrapper, {
    sdk: sdk
  }, entryElement));
}
let sdk;

// The application-level Sdk instance must be injected dynamically to avoid
// circular dependencies at the time of module resolution.
export function __injectSdkIntoInitializeBlock(_sdk) {
  sdk = _sdk;
}

// Exclusively for tests
export function __resetHasBeenInitialized() {
  hasBeenInitialized = false;
}
