// @airtable/blocks/dist/esm/shared/types/mutations_core.js
/** @module @airtable/blocks: mutations */ /** */

/** @hidden */
export const MutationTypesCore = Object.freeze({
  SET_MULTIPLE_GLOBAL_CONFIG_PATHS: 'setMultipleGlobalConfigPaths',
  SET_MULTIPLE_RECORDS_CELL_VALUES: 'setMultipleRecordsCellValues',
  DELETE_MULTIPLE_RECORDS: 'deleteMultipleRecords',
  CREATE_MULTIPLE_RECORDS: 'createMultipleRecords'
});

// /** @hidden */
// export type MutationType = ObjectValues<typeof MutationTypesCore>;

/**
 * The Mutation emitted when the App modifies one or more values in the
 * {@link GlobalConfig}.
 *
 * @docsPath testing/mutations/SetMultipleGlobalConfigPathsMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/SetMultipleRecordsCellValuesMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App deletes one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/DeleteMultipleRecordsMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App creates one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/CreateMultipleRecordsMutation
 */

/** @hidden */

/** @hidden */

/** @hidden */

/** */

/** */

/** Indicates whether the user has permission to perform a particular action, and if not, why. */
