// @airtable/blocks/dist/esm/base/models/record.js
/** @module @airtable/blocks/models: Record */ /** */

import { RecordCore, WatchableCellValueInFieldKeyPrefix, WatchableRecordKeysCore } from '../../shared/models/record_core';
import { isEnumValue } from '../../shared/private_utils';
import { invariant } from '../../shared/error_utils';
import colorUtils from '../../shared/color_utils';
import LinkedRecordsQueryResult from './linked_records_query_result';
import ObjectPool from './object_pool';
import RecordQueryResult from './record_query_result';
const WatchableRecordKeys = Object.freeze({
  ...WatchableRecordKeysCore,
  commentCount: 'commentCount'
});

// TODO: Should we load view data when this is watched?
// See comment on WatchableCellValueInFieldKeyPrefix.
const WatchableColorInViewKeyPrefix = 'colorInView:';
/**
 * Any key within record that can be watched:
 * - `'name'`
 * - `'commentCount'`
 * - `'cellValues'`
 * - `'cellValueInField:' + someFieldId`
 * - `'colorInView:' + someViewId`
 */

/**
 * Model class representing a record in a table.
 *
 * Do not instantiate. You can get instances of this class by calling `table.selectRecords`
 * or `view.selectRecords` and using the resulting {@link RecordQueryResult}.
 *
 * @docsPath models/Record
 */
class Record extends RecordCore {
  /** @internal */
  static _className = 'Record';
  /** @internal */
  static _isWatchableKey(key) {
    return isEnumValue(WatchableRecordKeys, key) || key.startsWith(WatchableCellValueInFieldKeyPrefix) || key.startsWith(WatchableColorInViewKeyPrefix);
  }
  /** @internal */

  /**
   * @internal
   */
  constructor(sdk, parentRecordStore, parentTable, recordId) {
    super(sdk, parentRecordStore, parentTable, recordId);
    this.__linkedRecordsQueryResultPool = new ObjectPool(LinkedRecordsQueryResult);
  }
  /**
   * @internal
   */
  _getViewMatching(viewOrViewIdOrViewName) {
    return this.parentTable.__getViewMatching(viewOrViewIdOrViewName);
  }

  /**
   * Gets the cell value of the given field for this record, formatted as a `string`.
   *
   * @param fieldOrFieldIdOrFieldName The field (or field ID or field name) whose cell value you'd like to get.
   * @example
   * ```js
   * const stringValue = myRecord.getCellValueAsString(myNumberField);
   * console.log(stringValue);
   * // => '42'
   * ```
   */
  getCellValueAsString(fieldOrFieldIdOrFieldName) {
    const field = this._getFieldMatching(fieldOrFieldIdOrFieldName);
    invariant(this._parentRecordStore.areCellValuesLoadedForFieldId(field.id), 'Cell values for field %s are not loaded', field.id);
    return super.getCellValueAsString(field.id);
  }
  _getRawCellValue(field) {
    invariant(this._parentRecordStore.areCellValuesLoadedForFieldId(field.id), 'Cell values for field %s are not loaded', field.id);
    return super._getRawCellValue(field);
  }

  /**
   * Gets the color of this record in a given view, or null if the record has no color in that
   * view.
   *
   * Can be watched with the 'colorInView:${ViewId}' key.
   *
   * @param viewOrViewIdOrViewName The view (or view ID or view name) to use for record coloring.
   */
  getColorInView(viewOrViewIdOrViewName) {
    const view = this._getViewMatching(viewOrViewIdOrViewName);
    return this._parentRecordStore.getViewDataStore(view.id).getRecordColor(this);
  }
  /**
   * Gets the CSS hex string for this record in a given view, or null if the record has no color
   * in that view.
   *
   * Can be watched with the 'colorInView:${ViewId}' key.
   *
   * @param viewOrViewIdOrViewName The view (or view ID or view name) to use for record coloring.
   */
  getColorHexInView(viewOrViewIdOrViewName) {
    const color = this.getColorInView(viewOrViewIdOrViewName);
    if (!color) {
      return null;
    }
    return colorUtils.getHexForColor(color);
  }
  /**
   * Select records referenced in a `multipleRecordLinks` cell value. Returns a query result
   * containing the records in the given `multipleRecordLinks` field.
   * See {@link RecordQueryResult} for more.
   *
   * @param fieldOrFieldIdOrFieldName The `multipleRecordLinks` field (or field ID or field name) to use.
   * @param opts Options for the query, such as sorts and fields.
   */
  selectLinkedRecordsFromCell(fieldOrFieldIdOrFieldName) {
    let opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    const field = this._getFieldMatching(fieldOrFieldIdOrFieldName);
    const linkedTableId = field.options && field.options.linkedTableId;
    invariant(typeof linkedTableId === 'string', 'linkedTableId must be set');
    const linkedTable = this._sdk.base.getTableById(linkedTableId);
    const linkedRecordStore = this._sdk.base.__getRecordStore(linkedTableId);
    const normalizedOpts = RecordQueryResult._normalizeOpts(linkedTable, linkedRecordStore, opts);
    return this.__linkedRecordsQueryResultPool.getObjectForReuse(this, field, normalizedOpts, this._sdk);
  }
  /**
   * Select and load records referenced in a `multipleRecordLinks` cell value. Returns a query result
   * promise containing the records in the given `multipleRecordLinks` field.
   * See {@link RecordQueryResult} for more.
   *
   * Remember to call `queryResult.unloadData` once you're finished with the query.
   *
   * @param fieldOrFieldIdOrFieldName The `multipleRecordLinks` field (or field ID or field name) to use.
   * @param opts Options for the query, such as sorts and fields.
   */
  async selectLinkedRecordsFromCellAsync(fieldOrFieldIdOrFieldName) {
    let opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    const queryResult = this.selectLinkedRecordsFromCell(fieldOrFieldIdOrFieldName, opts);
    await queryResult.loadDataAsync();
    return queryResult;
  }
  /**
   * The URL for the record. You can visit this URL in the browser to be taken to the record in the Airtable UI.
   *
   * @example
   * ```js
   * console.log(myRecord.url);
   * // => 'https://airtable.com/appxxxxxxxxxxxxxx/tblxxxxxxxxxxxxxx/recxxxxxxxxxxxxxx'
   * ```
   */
  get url() {
    return this._sdk.__airtableInterface.urlConstructor.getRecordUrl(this.id, this.parentTable.id);
  }
  /**
   * The number of comments on this record.
   *
   * @example
   * ```js
   * const commentCount = myRecord.commentCount;
   * const isSingular = commentCount === 1;
   * console.log(
   *     `This record has ${commentCount} comment${isSingular ? '' : 's'}`
   * );
   * ```
   */
  get commentCount() {
    return this._data.commentCount;
  }
  /**
   * @internal
   */
  __triggerOnChangeForDirtyPaths(dirtyPaths) {
    super.__triggerOnChangeForDirtyPaths(dirtyPaths);
    const {
      commentCount
    } = dirtyPaths;
    if (commentCount) {
      this._onChange(WatchableRecordKeys.commentCount);
    }
  }
  /**
   * @internal
   */
  __triggerOnChangeForRecordColorInViewId(viewId) {
    this._onChange(WatchableColorInViewKeyPrefix + viewId);
  }
}
export default Record;
