// @airtable/blocks/dist/esm/base/models/record_coloring.js
/** @module @airtable/blocks/models: Record Coloring */ /** */

// TODO(emma): update this to be an enum
/**
 * An enum of the different types of {@link recordColoring.modes}
 *
 * @hidden
 * @alias recordColoring.ModeTypes
 */
export const ModeTypes = {
  /**
   * @alias recordColoring.ModeTypes.NONE
   */
  NONE: 'none',
  /**
   * @alias recordColoring.ModeTypes.BY_SELECT_FIELD
   */
  BY_SELECT_FIELD: 'bySelectField',
  /**
   * @alias recordColoring.ModeTypes.BY_VIEW
   */
  BY_VIEW: 'byView'
};

/**
 * @hidden
 */

/** */

/** */

/** */

/** */

export const serialize = mode => {
  let result = mode.type;
  if (mode.type === ModeTypes.BY_SELECT_FIELD) {
    result += `-${mode.selectField.id}`;
  } else if (mode.type === ModeTypes.BY_VIEW) {
    result += `-${mode.view.id}`;
  }
  return result;
};

/**
 * Record coloring config creators.
 *
 * @alias recordColoring.modes
 * @example
 * ```js
 * import {recordColoring} from '@airtable/blocks/base/models';
 * import {useRecords} from '@airtable/blocks/base/ui';
 *
 * // no record coloring:
 * const recordColorMode = recordColoring.modes.none();
 * // color by select field:
 * const recordColorMode = recordColoring.modes.bySelectField(someSelectField);
 * // color from view:
 * const recordColorMode = recordColoring.modes.byView(someView);
 *
 * // with useRecords:
 * const queryResult = useRecords(table, { recordColorMode })
 *
 * // with a query result:
 * const queryResult = table.selectRecords({ recordColorMode });
 * ```
 */
export const modes = {
  /**
   * Returns a {@link RecordColorMode} that represents no record coloring.
   *
   * @alias recordColoring.modes.none
   */
  none() {
    return {
      type: ModeTypes.NONE
    };
  },
  /**
   * Returns a {@link RecordColorMode} that colors records by the specified Select field.
   *
   * @alias recordColoring.modes.bySelectField
   * @param selectField
   */
  bySelectField(selectField) {
    return {
      type: ModeTypes.BY_SELECT_FIELD,
      selectField
    };
  },
  /**
   * Returns a {@link RecordColorMode} that colors records by the specified View.
   *
   * @alias recordColoring.modes.byView
   * @param view
   */
  byView(view) {
    return {
      type: ModeTypes.BY_VIEW,
      view
    };
  }
};
