// @airtable/blocks/dist/esm/shared/models/mutations_core.js
import { MutationTypesCore } from '../types/mutations_core';
import { spawnError } from '../error_utils';
import { entries } from '../private_utils';
// Limit for how many items can be updated from a single batch mutation.
// This is number of records for MULTIPLE_RECORDS type mutations, and number of global config paths
// for SET_MULTIPLE_GLOBAL_CONFIG_PATHS.
// Same limit is enforced liveapp-side
export const MUTATIONS_MAX_BATCH_SIZE = 50;

// Liveapp requests must be under 2mb in size: we enforce a 1.9mb limit here to allow space for
// the other parts of the request
const MUTATIONS_MAX_BODY_SIZE = 1.9 * 1024 * 1024;
const MUTATION_HOLD_FOR_MS = 100;

/** @hidden */
export class MutationsCore {
  /** @internal */

  /** @internal */

  /** @internal */

  /** @internal */

  /** @internal */

  /** @internal */

  /** @hidden */
  constructor(sdk, session, base, applyModelChanges, applyGlobalConfigUpdates) {
    this._airtableInterface = sdk.__airtableInterface;
    this._session = session;
    this._sdk = sdk;
    this._base = base;
    this._applyModelChanges = applyModelChanges;
    this._applyGlobalConfigUpdates = applyGlobalConfigUpdates;
  }

  /** @hidden */
  async applyMutationAsync(mutation) {
    this._assertMutationIsValid(mutation);
    // Limit check is after validity check so that we display errors for when users pass in
    // objects correctly (eg updating linked records cell value to be a record object) -
    // otherwise the limit check will fail due to circular objects being converted to JSON first
    this._assertMutationUnderLimits(mutation);
    const permissionCheck = this.checkPermissionsForMutation(mutation);
    if (!permissionCheck.hasPermission) {
      throw spawnError('Cannot apply %s mutation: %s', mutation.type, permissionCheck.reasonDisplayString);
    }
    const didApplyOptimisticUpdates = this._applyOptimisticUpdatesForMutation(mutation);
    try {
      await this._getAirtableInterfaceAsAirtableInterfaceCore().applyMutationAsync(mutation, {
        holdForMs: MUTATION_HOLD_FOR_MS
      });
    } catch (err) {
      if (didApplyOptimisticUpdates) {
        // if we applied optimistic updates, we can't gracefully handle a promise rejection
        // here - we can't un-apply optimistic updates, so the SDK's internal data model is
        // in an unexpected state. Instead of letting this promise get rejected, throw an
        // error after an async gap to crash the block, and make this promise await
        // something that will never resolve so we don't run any of the developers error-
        // handling code.
        setTimeout(() => {
          throw err;
        }, 0);
        await new Promise(() => {});
      } else {
        throw err;
      }
    }
  }

  /** @hidden */
  checkPermissionsForMutation(mutation) {
    return this._getAirtableInterfaceAsAirtableInterfaceCore().checkPermissionsForMutation(mutation, this._base.__getBaseData());
  }

  // Not really sure why this is necessary. It seems like all callsites could just use
  // this._airtableInterface directly, but TS complains.
  /** @hidden */
  _getAirtableInterfaceAsAirtableInterfaceCore() {
    return this._airtableInterface;
  }

  /** @internal */
  _assertMutationUnderLimits(mutation) {
    // Two limits to check here:
    // - for record-related mutations, it isn't above MUTATIONS_MAX_BATCH_SIZE
    // - mutation payload size won't exceed liveapp request payload size limit
    // Requests are sent as form-encoded utf-8 (1 byte characters)
    if (encodeURIComponent(JSON.stringify(mutation)).length > MUTATIONS_MAX_BODY_SIZE) {
      throw spawnError('Request exceeds maximum size limit of %s bytes', MUTATIONS_MAX_BODY_SIZE);
    }
    if (this._doesMutationExceedBatchSizeLimit(mutation)) {
      throw spawnError('Request exceeds maximum batch size limit of %s items', MUTATIONS_MAX_BATCH_SIZE);
    }
  }

  /** @internal */
  _applyOptimisticUpdatesForMutation(mutation) {
    // GlobalConfig updates are different to other mutations (on models): for models, we
    // only apply optimistic updates if the relevant models are loaded, whereas for
    // SET_MULTIPLE_GLOBAL_CONFIG_PATHS we always apply optimistic updates.
    if (mutation.type === MutationTypesCore.SET_MULTIPLE_GLOBAL_CONFIG_PATHS) {
      // Not sure why this cast is necessary, but TS complains without it.
      const _mutation = mutation;
      this._applyGlobalConfigUpdates(_mutation.updates);
      return true;
    }
    const modelChanges = this._getOptimisticModelChangesForMutation(mutation);
    if (modelChanges.length > 0) {
      this._applyModelChanges(modelChanges);
      return true;
    }
    return false;
  }

  /** @internal */
  _doesMutationExceedBatchSizeLimit(mutation) {
    switch (mutation.type) {
      case MutationTypesCore.SET_MULTIPLE_RECORDS_CELL_VALUES:
      case MutationTypesCore.CREATE_MULTIPLE_RECORDS:
        return mutation.records.length > MUTATIONS_MAX_BATCH_SIZE;
      case MutationTypesCore.DELETE_MULTIPLE_RECORDS:
        return mutation.recordIds.length > MUTATIONS_MAX_BATCH_SIZE;
      case MutationTypesCore.SET_MULTIPLE_GLOBAL_CONFIG_PATHS:
        return mutation.updates.length > MUTATIONS_MAX_BATCH_SIZE;
      // The "default" case cannot be reached because this method is only
      // invoked after the mutation type has been validated.
      // istanbul ignore next
      default:
        return false;
    }
  }

  /** @internal */
  _assertFieldIsValidForMutation(field) {
    if (field.isComputed) {
      throw spawnError("Can't set cell values: Field '%s' is computed and cannot be set", field.name);
    }
  }

  /** @internal */
  _assertMutationIsValid(mutation) {
    // We call validate the data (including any cell values) because if the data required for
    // us to do that is already loaded in the block, we can error out before applying
    // optimistic updates or sending the update to liveapp. That means that the error is
    // recoverable. Once we apply optimistic updates, if liveapp rejects the update then we
    // can't recover from that - we have to crash the block. We _could_ skip over these
    // validations - it wouldn't cause issues outside of the block frame. But running them
    // gives us slightly more confidence that we can do something other than completely crash
    // the block in the event of an invalid mutation.

    const appInterface = this._sdk.__appInterface;
    switch (mutation.type) {
      case MutationTypesCore.SET_MULTIPLE_RECORDS_CELL_VALUES:
        {
          const {
            tableId,
            records
          } = mutation;
          const table = this._base.getTableByIdIfExists(tableId);
          if (!table) {
            throw spawnError("Can't set cell values: No table with id %s exists", tableId);
          }

          // For every mutation, we check that we're not trying to set fields that we don't support
          // mutations for yet. When record data is loaded, we also check that the record we're
          // updating actually exists and that the cell values are valid.
          const recordStore = this._base.__getRecordStore(tableId);
          const checkedFieldIds = new Set();
          for (const record of records) {
            let existingRecord = null;
            if (this._isRecordStoreReadyForMutations(recordStore)) {
              existingRecord = recordStore.getRecordByIdIfExists(record.id);
              if (!existingRecord) {
                throw spawnError("Can't set cell values: No record with id %s exists", record.id);
              }
            }
            for (const fieldId of Object.keys(record.cellValuesByFieldId)) {
              const field = table.getFieldByIdIfExists(fieldId);
              if (!field) {
                throw spawnError("Can't set cell values: No field with id %s exists in table '%s'", fieldId, table.name);
              }
              if (!checkedFieldIds.has(fieldId)) {
                this._assertFieldIsValidForMutation(field);
                checkedFieldIds.add(fieldId);
              }
              if (existingRecord && this._isFieldAvailableForMutation(recordStore, field.id)) {
                const validationResult = this._airtableInterface.fieldTypeProvider.validateCellValueForUpdate(appInterface, record.cellValuesByFieldId[fieldId], existingRecord._getRawCellValue(field), field._data);
                if (!validationResult.isValid) {
                  throw spawnError("Can't set cell values: invalid cell value for field '%s'.\n%s", field.name, validationResult.reason);
                }
              }
            }
          }
          return;
        }
      case MutationTypesCore.DELETE_MULTIPLE_RECORDS:
        {
          const {
            tableId,
            recordIds
          } = mutation;
          const table = this._base.getTableByIdIfExists(tableId);
          if (!table) {
            throw spawnError("Can't delete records: No table with id %s exists", tableId);
          }
          const recordStore = this._base.__getRecordStore(tableId);
          if (this._isRecordStoreReadyForMutations(recordStore)) {
            for (const recordId of recordIds) {
              const record = recordStore.getRecordByIdIfExists(recordId);
              if (!record) {
                throw spawnError("Can't delete records: No record with id %s exists in table '%s'", recordId, table.name);
              }
            }
          }
          return;
        }
      case MutationTypesCore.CREATE_MULTIPLE_RECORDS:
        {
          const {
            tableId,
            records
          } = mutation;
          const checkedFieldIds = new Set();
          const table = this._base.getTableByIdIfExists(tableId);
          if (!table) {
            throw spawnError("Can't create records: No table with id %s exists", tableId);
          }
          for (const record of records) {
            for (const fieldId of Object.keys(record.cellValuesByFieldId)) {
              const field = table.getFieldByIdIfExists(fieldId);
              if (!field) {
                throw spawnError("Can't create records: No field with id %s exists in table '%s'", fieldId, table.name);
              }
              if (!checkedFieldIds.has(fieldId)) {
                this._assertFieldIsValidForMutation(field);
                checkedFieldIds.add(fieldId);
              }

              // Current cell value is null since the record doesn't exist.
              const validationResult = this._airtableInterface.fieldTypeProvider.validateCellValueForUpdate(appInterface, record.cellValuesByFieldId[fieldId], null, field._data);
              if (!validationResult.isValid) {
                throw spawnError("Can't create records: invalid cell value for field '%s'.\n%s", field.name, validationResult.reason);
              }
            }
          }
          return;
        }
      case MutationTypesCore.SET_MULTIPLE_GLOBAL_CONFIG_PATHS:
        {
          // globalConfig update is a special case: globalConfig handles validation before
          // invoking this mutation, since it relies on internal state to validate the
          // paths being set.
          return;
        }
      default:
        throw spawnError('unhandled mutation type: %s', mutation.type);
    }
  }

  /** @internal */
  _getOptimisticModelChangesForMutation(mutation) {
    switch (mutation.type) {
      case MutationTypesCore.SET_MULTIPLE_RECORDS_CELL_VALUES:
        {
          const {
            tableId,
            records
          } = mutation;
          const recordStore = this._base.__getRecordStore(tableId);
          return records.flatMap(record => Object.keys(record.cellValuesByFieldId).filter(fieldId => this._isFieldAvailableForMutation(recordStore, fieldId)).map(fieldId => ({
            path: ['tablesById', tableId, 'recordsById', record.id, 'cellValuesByFieldId', fieldId],
            value: record.cellValuesByFieldId[fieldId]
          })));
        }
      case MutationTypesCore.DELETE_MULTIPLE_RECORDS:
        {
          const {
            tableId,
            recordIds
          } = mutation;
          return recordIds.map(recordId => ({
            path: ['tablesById', tableId, 'recordsById', recordId],
            value: undefined
          }));
        }
      case MutationTypesCore.CREATE_MULTIPLE_RECORDS:
        {
          const {
            tableId,
            records
          } = mutation;
          const recordStore = this._base.__getRecordStore(tableId);
          return records.map(record => {
            // Only apply optimistic changes for fields that are "available" for mutation
            const filteredCellValuesByFieldId = {};
            for (const [fieldId, cellValue] of entries(record.cellValuesByFieldId)) {
              if (this._isFieldAvailableForMutation(recordStore, fieldId)) {
                filteredCellValuesByFieldId[fieldId] = cellValue;
              }
            }
            return {
              path: ['tablesById', tableId, 'recordsById', record.id],
              value: {
                ...this._getDefaultRecordProperties(),
                id: record.id,
                cellValuesByFieldId: filteredCellValuesByFieldId
              }
            };
          });
        }

      // The following branch is unreachable because this method's only
      // call site is preceded by an explicit guard for this condition.
      // istanbul ignore next
      case MutationTypesCore.SET_MULTIPLE_GLOBAL_CONFIG_PATHS:
        {
          throw spawnError('attempting to generate model updates for SET_MULTIPLE_GLOBAL_CONFIG_PATH');
        }

      // The following branch is unreachable because this method's only
      // call site is preceded by an explicit guard for this condition.
      // istanbul ignore next
      default:
        throw spawnError('unhandled mutation type: %s', mutation.type);
    }
  }

  /** @internal */

  /** @internal */

  /** @internal */
}
