// @airtable/blocks/dist/esm/shared/models/record_core.js
import { cloneDeep, isObjectEmpty } from '../private_utils';
import { invariant } from '../error_utils';
import { FieldType } from '../types/field_core';
import AbstractModel from './abstract_model';
import { FieldCore } from './field_core';
export const WatchableRecordKeysCore = Object.freeze({
  name: 'name',
  // TODO(kasra): these keys don't have matching getters (not that they should
  // it's just inconsistent...)
  cellValues: 'cellValues'
});
// TODO: load cell values in field when this is watched? This will
// cause the CellRenderer component to load cell values, which seems okay,
// but needs a little more thought.
export const WatchableCellValueInFieldKeyPrefix = 'cellValueInField:';

/** @hidden */

/** @hidden */
export class RecordCore extends AbstractModel {
  /** @internal */
  static _className = 'RecordCore';
  /** @internal */

  /** @internal */

  /**
   * @internal
   */
  constructor(sdk, parentRecordStore, parentTable, recordId) {
    super(sdk, recordId);
    this._parentRecordStore = parentRecordStore;
    this._parentTable = parentTable;
  }

  /**
   * @internal
   */
  get _dataOrNullIfDeleted() {
    const tableData = this._baseData.tablesById[this.parentTable.id];
    if (!tableData) {
      return null;
    }
    const recordsById = tableData.recordsById;
    invariant(recordsById, 'Record data is not loaded');
    return recordsById[this._id] ?? null;
  }
  /**
   * The primary cell value in this record, formatted as a `string`.
   *
   * @example
   * ```js
   * console.log(myRecord.name);
   * // => '42'
   * ```
   */
  get name() {
    return this.getCellValueAsString(this.parentTable.primaryField);
  }
  /**
   * The created time of this record.
   *
   * @example
   * ```js
   * console.log(`
   *     This record was created at ${myRecord.createdTime.toISOString()}
   * `);
   * ```
   */
  get createdTime() {
    return new Date(this._data.createdTime);
  }

  /**
   * The table that this record belongs to. Should never change because records aren't moved between tables.
   *
   * @internal (since we may not be able to return parent model instances in the immutable models world)
   * @example
   * ```js
   * import {useRecords} from '@airtable/blocks/base/ui';
   * const records = useRecords(myTable);
   * console.log(records[0].parentTable.id === myTable.id);
   * // => true
   * ```
   */
  get parentTable() {
    return this._parentTable;
  }
  /**
   * @internal
   */
  _getFieldMatching(fieldOrFieldIdOrFieldName) {
    if (fieldOrFieldIdOrFieldName instanceof FieldCore) {
      return this._parentTable.__getFieldMatching(fieldOrFieldIdOrFieldName.id);
    }
    return this.parentTable.__getFieldMatching(fieldOrFieldIdOrFieldName);
  }

  /**
   * @internal
   *
   * For use when we need the raw public API cell value. Specifically makes a difference
   * for lookup fields, where we translate the format to a blocks-specific format in getCellValue.
   * That format is incompatible with fieldTypeProvider methods, which expect the public API
   * format - use _getRawCellValue instead.
   */
  _getRawCellValue(field) {
    const {
      cellValuesByFieldId
    } = this._data;
    if (!cellValuesByFieldId) {
      return null;
    }
    const cellValue = cellValuesByFieldId[field.id] !== undefined ? cellValuesByFieldId[field.id] : null;
    if (typeof cellValue === 'object' && cellValue !== null) {
      // Copy non-primitives.
      // TODO(kasra): maybe freezeDeep instead?
      return cloneDeep(cellValue);
    } else {
      return cellValue;
    }
  }
  /**
   * Gets the cell value of the given field for this record.
   *
   * @param fieldOrFieldIdOrFieldName The field (or field ID or field name) whose cell value you'd like to get.
   * @example
   * ```js
   * const cellValue = myRecord.getCellValue(mySingleLineTextField);
   * console.log(cellValue);
   * // => 'cell value'
   * ```
   */
  getCellValue(fieldOrFieldIdOrFieldName) {
    const field = this._getFieldMatching(fieldOrFieldIdOrFieldName);
    const cellValue = this._getRawCellValue(field);

    // HACK: migrate to new public lookup cell value format if needed
    if (typeof cellValue === 'object' && cellValue !== null && field.type === FieldType.MULTIPLE_LOOKUP_VALUES && !this._sdk.__airtableInterface.sdkInitData.isUsingNewLookupCellValueFormat) {
      const cellValueForMigration = [];
      invariant(Array.isArray(cellValue.linkedRecordIds), 'linkedRecordIds');
      for (const linkedRecordId of cellValue.linkedRecordIds) {
        invariant(typeof linkedRecordId === 'string', 'linkedRecordId');
        const {
          valuesByLinkedRecordId
        } = cellValue;
        invariant(valuesByLinkedRecordId && typeof valuesByLinkedRecordId === 'object', 'valuesByLinkedRecordId');
        const value = valuesByLinkedRecordId[linkedRecordId];
        if (Array.isArray(value)) {
          for (const v of value) {
            cellValueForMigration.push({
              linkedRecordId,
              value: v
            });
          }
        } else {
          cellValueForMigration.push({
            linkedRecordId,
            value
          });
        }
      }
      return cellValueForMigration;
    }
    return cellValue;
  }
  /**
   * Gets the cell value of the given field for this record, formatted as a `string`.
   *
   * @param fieldOrFieldIdOrFieldName The field (or field ID or field name) whose cell value you'd like to get.
   * @example
   * ```js
   * const stringValue = myRecord.getCellValueAsString(myNumberField);
   * console.log(stringValue);
   * // => '42'
   * ```
   */
  getCellValueAsString(fieldOrFieldIdOrFieldName) {
    const field = this._getFieldMatching(fieldOrFieldIdOrFieldName);
    const cellValue = this._getRawCellValue(field);
    if (cellValue === null || cellValue === undefined) {
      return '';
    } else {
      const airtableInterface = this._sdk.__airtableInterface;
      const appInterface = this._sdk.__appInterface;
      return airtableInterface.fieldTypeProvider.convertCellValueToString(appInterface, cellValue, field._data);
    }
  }
  /**
   * Returns a URL that is suitable for rendering an attachment on the current client.
   * The URL that is returned will only work for the current user.
   *
   * @param attachmentId The ID of the attachment.
   * @param attachmentUrl The attachment's URL (which is not suitable for rendering on the client).
   * @example
   * ```js
   * import React from 'react';
   *
   * function RecordAttachments(props) {
   *     const {record, attachmentField} = props;
   *     const attachmentCellValue = record.getCellValue(attachmentField);
   *     if (attachmentCellValue === null) {
   *         return null;
   *     }
   *     return (
   *         <div>
   *             {attachmentCellValue.map(attachmentObj => {
   *                 const clientUrl =
   *                     record.getAttachmentClientUrlFromCellValueUrl(
   *                         attachmentObj.id,
   *                         attachmentObj.url
   *                     );
   *                 return (
   *                     <img key={attachmentObj.id} src={clientUrl} width={200} />
   *                 );
   *             })}
   *         </div>
   *     );
   * }
   * ```
   */
  getAttachmentClientUrlFromCellValueUrl(attachmentId, attachmentUrl) {
    const airtableInterface = this._sdk.__airtableInterface;
    const appInterface = this._sdk.__appInterface;
    return airtableInterface.urlConstructor.getAttachmentClientUrl(appInterface, attachmentId, attachmentUrl);
  }
  /**
   * @internal
   */
  __triggerOnChangeForDirtyPaths(dirtyPaths) {
    const {
      cellValuesByFieldId
    } = dirtyPaths;
    if (cellValuesByFieldId && !isObjectEmpty(cellValuesByFieldId)) {
      // TODO: don't trigger changes for fields that aren't supposed to be loaded
      // (in some cases, e.g. record created, liveapp will send cell values
      // that we're not subscribed to).

      this._onChange(WatchableRecordKeysCore.cellValues, Object.keys(cellValuesByFieldId));
      for (const fieldId of Object.keys(cellValuesByFieldId)) {
        this._onChange(WatchableCellValueInFieldKeyPrefix + fieldId, fieldId);
      }
      if (cellValuesByFieldId[this.parentTable.primaryField.id]) {
        this._onChange(WatchableRecordKeysCore.name);
      }
    }
  }
}
