// @airtable/blocks/dist/esm/shared/models/record_store_core.js
import { isEnumValue, entries, has } from '../../shared/private_utils';
import { invariant } from '../../shared/error_utils';
import AbstractModel from '../../shared/models/abstract_model';
export const WatchableRecordStoreKeysCore = Object.freeze({
  records: 'records',
  recordIds: 'recordIds',
  cellValues: 'cellValues'
});
export const WatchableCellValuesInFieldKeyPrefix = 'cellValuesInField:';

/**
 * The string case is to accommodate prefix keys
 *
 * @internal
 */

/**
 * One RecordStore exists per table, and contains all the record data associated with that table.
 * Table itself is for schema information only, so isn't the appropriate place for this data.
 *
 * @internal
 */
class RecordStoreCore extends AbstractModel {
  static _className = 'RecordStoreCore';
  static _isWatchableKey(key) {
    return isEnumValue(WatchableRecordStoreKeysCore, key) || key.startsWith(WatchableCellValuesInFieldKeyPrefix);
  }
  _recordModelsById = {};
  constructor(sdk, tableId) {
    super(sdk, `${tableId}-RecordStore`);
    this.tableId = tableId;
  }
  get _dataOrNullIfDeleted() {
    return this._baseData.tablesById[this.tableId] ?? null;
  }
  watch(keys, callback, context) {
    const validKeys = super.watch(keys, callback, context);
    return validKeys;
  }
  unwatch(keys, callback, context) {
    const validKeys = super.unwatch(keys, callback, context);
    return validKeys;
  }

  /**
   * The records in this table.
   */
  get records() {
    const recordsById = this._data.recordsById;
    invariant(recordsById, 'Record metadata is not loaded');
    const records = this.recordIds.map(recordId => {
      const record = this.getRecordByIdIfExists(recordId);
      invariant(record, 'record not found');
      return record;
    });
    return records;
  }

  /**
   * The record IDs in this table.
   */

  getRecordByIdIfExists(recordId) {
    const recordsById = this._data.recordsById;
    invariant(recordsById, 'Record metadata is not loaded');
    invariant(typeof recordId === 'string', 'getRecordById expects a string');
    if (!recordsById[recordId]) {
      return null;
    } else {
      if (this._recordModelsById[recordId]) {
        return this._recordModelsById[recordId];
      }
      const newRecord = this._constructRecord(recordId, this._sdk.base.getTableById(this.tableId));
      this._recordModelsById[recordId] = newRecord;
      return newRecord;
    }
  }
  triggerOnChangeForDirtyPaths(dirtyPaths) {
    if (dirtyPaths.recordsById) {
      // Since tables don't have a record order, need to detect if a record
      // was created or deleted and trigger onChange for records.
      const dirtyFieldIdsSet = {};
      const addedRecordIds = [];
      const removedRecordIds = [];
      for (const [recordId, dirtyRecordPaths] of entries(dirtyPaths.recordsById)) {
        if (dirtyRecordPaths && dirtyRecordPaths._isDirty) {
          // If the entire record is dirty, it was either created or deleted.
          invariant(this._data.recordsById, 'No recordsById');
          if (has(this._data.recordsById, recordId)) {
            addedRecordIds.push(recordId);
          } else {
            removedRecordIds.push(recordId);
            const recordModel = this._recordModelsById[recordId];
            if (recordModel) {
              // Remove the Record model if it was deleted.
              delete this._recordModelsById[recordId];
            }
          }
        } else {
          const recordModel = this._recordModelsById[recordId];
          if (recordModel) {
            recordModel.__triggerOnChangeForDirtyPaths(dirtyRecordPaths);
          }
        }
        const {
          cellValuesByFieldId
        } = dirtyRecordPaths;
        if (cellValuesByFieldId) {
          for (const fieldId of Object.keys(cellValuesByFieldId)) {
            dirtyFieldIdsSet[fieldId] = true;
          }
        }
      }

      // Now that we've composed our created/deleted record ids arrays, let's fire
      // the records onChange event if any records were created or deleted.
      if (addedRecordIds.length > 0 || removedRecordIds.length > 0) {
        this._onChange(WatchableRecordStoreKeysCore.records, {
          addedRecordIds,
          removedRecordIds
        });
        this._onChange(WatchableRecordStoreKeysCore.recordIds, {
          addedRecordIds,
          removedRecordIds
        });
      }

      // NOTE: this is an experimental (and somewhat messy) way to watch
      // for changes to cells in a table, as an alternative to implementing
      // full event bubbling. For now, it unblocks the things we want to
      // build, but we may replace it.
      // If we keep it, could be more efficient by not calling _onChange
      // if there are no subscribers.
      // TODO: don't trigger changes for fields that aren't supposed to be loaded
      // (in some cases, e.g. record created, liveapp will send cell values
      // that we're not subscribed to).
      const fieldIds = Object.freeze(Object.keys(dirtyFieldIdsSet));
      const recordIds = Object.freeze(Object.keys(dirtyPaths.recordsById));
      if (fieldIds.length > 0 && recordIds.length > 0) {
        this._onChange(WatchableRecordStoreKeysCore.cellValues, {
          recordIds,
          fieldIds
        });
      }
      for (const fieldId of fieldIds) {
        this._onChange(WatchableCellValuesInFieldKeyPrefix + fieldId, recordIds, fieldId);
      }
    }
  }
}

/** @internal */
export default RecordStoreCore;
