// @airtable/blocks/dist/esm/testing/interface/abstract_mock_airtable_interface.js
// istanbul ignore file
import EventEmitter from 'events';
import { cloneDeep } from '../../shared/private_utils';
import { spawnError } from '../../shared/error_utils';
/** @internal */
const fieldTypeProvider = {
  isComputed(fieldData) {
    return false;
  },
  getConfig: (appInterface, fieldData, fieldNamesById) => {
    return {
      type: fieldData.type,
      options: fieldData.typeOptions
    };
  },
  convertStringToCellValue(appInterface, string, fieldData) {
    return '';
  },
  convertCellValueToString(appInterface, cellValue, fieldData) {
    return '';
  },
  getCellRendererData(appInterface, cellValue, fieldData, shouldWrap) {
    return {
      cellValueHtml: `<pre>${JSON.stringify(cellValue)}</pre>`,
      attributes: {}
    };
  },
  validateCellValueForUpdate(appInterface, newCellValue, currentCellValue, fieldData) {
    return {
      isValid: true
    };
  }
};

/** @internal */
const urlConstructor = {
  getAttachmentClientUrl(appInterface, attachmentId, attachmentUrl) {
    return attachmentUrl;
  }
};

/** @internal */
const globalConfigHelpers = {
  validatePath(path, store) {
    return {
      isValid: true
    };
  },
  validateAndApplyUpdates(updates, store) {
    throw spawnError('validateAndApplyUpdates unimplemented');
  }
};

/** @internal */
const idGenerator = {
  generateRecordId: () => 'recGeneratedMockId'
};

/**
 * An abstract base class with a common interface exposed to both Blocks SDK's
 * internal automated test suite and the blocks-testing public repo.
 *
 * @hidden
 */
export class AbstractMockAirtableInterface extends EventEmitter {
  constructor(initData) {
    super();
    this._initData = cloneDeep(initData);
    this.reset();
  }

  /**
   * Revert the mock interface to its initial state. This includes:
   *
   * - removing all event listeners
   * - restoring the database schema
   */
  reset() {
    this.removeAllListeners();
    this.sdkInitData = cloneDeep(this._initData);
  }
  get fieldTypeProvider() {
    return fieldTypeProvider;
  }
  get urlConstructor() {
    return urlConstructor;
  }
  get globalConfigHelpers() {
    return globalConfigHelpers;
  }
  get idGenerator() {
    return idGenerator;
  }
  assertAllowedSdkPackageVersion() {}
  applyMutationAsync(mutation, opts) {
    return Promise.resolve();
  }
  checkPermissionsForMutation(mutation) {
    return {
      hasPermission: true
    };
  }
  fetchForeignRecordsAsync(tableId, recordId, fieldId, filterString) {
    return Promise.resolve({
      records: []
    });
  }
  setCustomPropertiesAsync(properties) {
    return Promise.resolve(true);
  }
  setSelectedSubElementAsync(selectedSubElement) {
    return Promise.resolve(true);
  }
  fetchAndSubscribeToSelectionDataAsync(callback) {
    this.on('selectedSubElementIdUpdates', callback);
    return Promise.resolve({
      selectedSubElementId: null
    });
  }
  unsubscribeFromSelectionData() {}
  subscribeToModelUpdates(fn) {
    this.on('modelupdates', fn);
  }

  // TODO(jugglinmike): Implement
  subscribeToGlobalConfigUpdates() {}
  triggerModelUpdates(changes) {
    this.emit('modelupdates', {
      changes
    });
  }

  // TODO(jugglinmike): Implement
  triggerGlobalConfigUpdates() {}
}
