// @airtable/blocks/dist/esm/shared/color_utils.js
/** @module @airtable/blocks/ui: colorUtils */ /** */
import { getEnumValueIfExists, has } from './private_utils';
import Colors, { rgbTuplesByColor } from './colors';

/** A red/green/blue color object. Each property is a number from 0 to 255. */

/**
 * Utilities for working with {@link Color} names from the {@link colors} enum.
 *
 * @docsPath UI/utils/colorUtils
 */

const colorUtils = {
  getHexForColor: colorString => {
    const color = getEnumValueIfExists(Colors, colorString);
    if (!color) {
      return null;
    }
    const rgbTuple = rgbTuplesByColor[color];
    const hexNumber = rgbTuple[0] << 16 | rgbTuple[1] << 8 | rgbTuple[2];
    return `#${hexNumber.toString(16).padStart(6, '0')}`;
  },
  getRgbForColor: colorString => {
    const color = getEnumValueIfExists(Colors, colorString);
    if (!color) {
      return null;
    }
    const rgbTuple = rgbTuplesByColor[color];
    return {
      r: rgbTuple[0],
      g: rgbTuple[1],
      b: rgbTuple[2]
    };
  },
  shouldUseLightTextOnColor: colorString => {
    if (!has(rgbTuplesByColor, colorString)) {
      // Don't have a color for this. Let's just return false as a default
      // instead of throwing.
      return false;
    }

    // Light1 and Light2 colors use dark text.
    // Bright, Dark1 and no suffix colors use light text.
    // NOTE: use shouldUseDarkText instead of shouldUseLightText just to make
    // checking the suffix easier, since no suffix uses light text.
    const shouldUseDarkText = colorString.endsWith('Light1') || colorString.endsWith('Light2');
    return !shouldUseDarkText;
  }
};
export default colorUtils;
