// @airtable/blocks/dist/esm/shared/private_utils.js
import getAirtableInterface from '../injected/airtable_interface';
import { spawnError } from './error_utils';

/** @hidden */

/** @hidden */

/** @hidden */

/**
 * Get the type of all the values of an object.
 *
 * Same as the legacy Flow `$Values<T>` type.
 *
 * @hidden
 */

/** @hidden */

/**
 * Allows creating an object map type with a dynamic key type.
 *
 * TypeScript only allows `string` for `K` in `{[key: K]: V}` so we need a utility to bridge
 * the gap.
 *
 * This is an alias for TypeScript’s `Record` type, but the name “record” is confusing given our
 * Airtable domain model.
 *
 * @hidden
 */

/**
 * Creates an enum from provided string arguments.
 *
 * Useful for consumer-facing enums (eg `Button#variant`) where we want to make the external
 * developer experience convenient by providing a string value, but also want to internally
 * reference enum values using object notation.
 *
 * @hidden
 */
export function createEnum() {
  const spec = {};
  for (var _len = arguments.length, enumValues = new Array(_len), _key = 0; _key < _len; _key++) {
    enumValues[_key] = arguments[_key];
  }
  for (const value of enumValues) {
    spec[value] = value;
  }
  return Object.freeze(spec);
}

/**
 * Creates a Type for an enum created using `createEnum`.
 *
 * @hidden
 */

/**
 * Safely cast a value to the type passed in as a type parameter.
 *
 * This was added in the TypeScript migration to provide a safe translation for Flow’s type cast
 * operator `(x: T)`.
 *
 * @hidden
 */
export function cast(x) {
  return x;
}

/**
 * @hidden
 */
export function cloneDeep(obj) {
  const jsonString = JSON.stringify(obj);
  if (jsonString === undefined) {
    return obj;
  }
  return JSON.parse(jsonString);
}

/**
 * This is basically taken from https://github.com/epoberezkin/fast-deep-equal/blob/a8e7172b6c411ec320d6045fd4afbd2abc1b4bde/src/index.jst
 * with some minor adjustments based on our needs.
 * Differences:
 * - Removed RegExp support because we don't need it
 * - Removed .valueOf() and .toString() support because we don't need it
 */
export function isDeepEqual(a, b) {
  if (a === b) {
    return true;
  }
  if (a && b && typeof a === 'object' && typeof b === 'object') {
    if (a.constructor !== b.constructor) {
      return false;
    }
    if (Array.isArray(a)) {
      const length = a.length;
      if (length !== b.length) {
        return false;
      }
      for (let i = length - 1; i >= 0; i--) {
        if (!isDeepEqual(a[i], b[i])) {
          return false;
        }
      }
      return true;
    }
    const keys = Object.keys(a);
    const length = keys.length;
    if (length !== Object.keys(b).length) {
      return false;
    }
    for (let i = length - 1; i >= 0; i--) {
      if (!Object.prototype.hasOwnProperty.call(b, keys[i])) {
        return false;
      }
    }
    for (let i = length - 1; i >= 0; i--) {
      const key = keys[i];
      if (!isDeepEqual(a[key], b[key])) {
        return false;
      }
    }
    return true;
  }

  // true if both NaN, false otherwise
  // eslint-disable-next-line no-self-compare
  return a !== a && b !== b;
}

// typescript has a stricter definition for Object.values and Object.entries that return mixed in place
// of the actual values. This is because for non-exact objects, that's the only sound definition.
// You can call Object.values with a value typed as {x: number} that actually looks like
// {x: number, y: string}, for example. Returning mixed isn't particularly useful though, so we
// provide these unsound wrappers instead.
// TODO: consider renaming these with unsound_ prefixes.
/**
 * @hidden
 */
export function keys(obj) {
  return Object.keys(obj);
}

/**
 * @hidden
 */
export function values(obj) {
  return Object.values(obj);
}

/**
 * @hidden
 */
export function entries(obj) {
  // @ts-ignore
  return Object.entries(obj);
}

// Result values are discarded and errors are thrown asynchronously.
// NOTE: this is different from the one in u: the function passed
// in must be fully bound with all of its arguments and will be immediately
// called (this does not return a function). This makes it work better with
// Flow: you get argument type checking by using `.bind`.
/**
 * @hidden
 */
export function fireAndForgetPromise(fn) {
  fn().catch(err => {
    // Defer til later, so the error doesn't cause the promise to be rejected.
    setTimeout(() => {
      throw err;
    }, 0);
  });
}

/**
 * @hidden
 */
export function has(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}
const invertedEnumCache = new WeakMap();
/**
 * @hidden
 */
function getInvertedEnumMemoized(enumObj) {
  const existingInvertedEnum = invertedEnumCache.get(enumObj);
  if (existingInvertedEnum) {
    // typescript can't type WeakMap precisely enough to know that it's being used as this sort of cache
    return existingInvertedEnum;
  }
  const invertedEnum = {};
  for (const enumKey of keys(enumObj)) {
    const enumValue = enumObj[enumKey];
    invertedEnum[enumValue] = enumKey;
  }
  invertedEnumCache.set(enumObj, invertedEnum);
  return invertedEnum;
}

/**
 * @hidden
 */
export function getEnumValueIfExists(enumObj, valueToCheck) {
  const invertedEnum = getInvertedEnumMemoized(enumObj);
  if (has(invertedEnum, valueToCheck) && invertedEnum[valueToCheck]) {
    const enumKey = invertedEnum[valueToCheck];
    return enumObj[enumKey];
  }
  return null;
}

/**
 * @hidden
 */
export function assertEnumValue(enumObj, valueToCheck) {
  const enumValue = getEnumValueIfExists(enumObj, valueToCheck);
  if (!enumValue) {
    throw spawnError('Unknown enum value %s', valueToCheck);
  }
  return enumValue;
}

/**
 * @hidden
 */
export function isEnumValue(enumObj, valueToCheck) {
  return getEnumValueIfExists(enumObj, valueToCheck) !== null;
}

/**
 * @hidden
 */
export function isObjectEmpty(obj) {
  for (const key in obj) {
    if (has(obj, key)) {
      return false;
    }
  }
  return true;
}

/**
 * @hidden
 */
export function isNullOrUndefinedOrEmpty(value) {
  return value === null || value === undefined || (typeof value === 'string' || Array.isArray(value)) && value.length === 0 || typeof value === 'object' && value !== null && isObjectEmpty(value);
}

/**
 * @hidden
 */
export function compact(array) {
  const result = [];
  for (const item of array) {
    if (item !== null && item !== undefined) {
      result.push(item);
    }
  }
  return result;
}

/**
 * @hidden
 */
export function clamp(n, lowerBound, upperBound) {
  if (n < lowerBound) {
    return lowerBound;
  } else if (n > upperBound) {
    return upperBound;
  } else {
    return n;
  }
}

/** @hidden */

/**
 * @hidden
 */
export function flattenDeep(array) {
  // @ts-ignore
  return array.flat(Infinity);
}

// For cases where the object keys aren't strings, it's the consumer's responsibility
// to convert them to strings in getKey, e.g. keyBy(collection, o => String(o.id))
/**
 * @hidden
 */
export function keyBy(array, getKey) {
  const result = {};
  for (const item of array) {
    result[getKey(item)] = item;
  }
  return result;
}

/**
 * @hidden
 */
export function getId(_ref) {
  let {
    id
  } = _ref;
  return id;
}

/**
 * @hidden
 */
export function uniqBy(array, getKey) {
  const usedKeys = new Set();
  const result = [];
  for (const item of array) {
    const key = getKey(item);
    if (!usedKeys.has(key)) {
      result.push(item);
      usedKeys.add(key);
    }
  }
  return result;
}
let idCount = 0;
/** @hidden */
export function getLocallyUniqueId() {
  let prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return `${prefix}.${idCount++}`;
}
const plainObjectPrototype = Object.getPrototypeOf({});
/**
 * A more restrictive version of Lodash's `get`. Notable differences:
 * - Will only search an object's own properties
 * - Only allows indexing into plain objects - searching in `number`, `string`, `Array`, `null`, or non-plain objects will throw
 *
 * @hidden
 */
export function getValueAtOwnPath(value, path) {
  let currentValue = value;
  for (const part of path) {
    if (currentValue === undefined) {
      return undefined;
    }
    if (typeof currentValue !== 'object' || currentValue === null) {
      throw spawnError("Cannot get '%s' in primitive value", part);
    }
    if (Array.isArray(currentValue)) {
      throw spawnError("Cannot get '%s' in array", part);
    }
    const prototype = Object.getPrototypeOf(currentValue);
    if (prototype !== null && prototype !== plainObjectPrototype) {
      throw spawnError("Cannot get '%s' in non-plain object", part);
    }
    currentValue = has(currentValue, part) ? currentValue[part] : undefined;
  }
  return currentValue;
}

/** @hidden */
export function arrayDifference(a, b) {
  const bSet = new Set(b);
  return a.filter(item => !bSet.has(item));
}

/** @hidden */
export function debounce(fn, timeoutMs) {
  let lastTimeoutId = null;
  return function () {
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    if (lastTimeoutId !== null) {
      clearTimeout(lastTimeoutId);
    }
    lastTimeoutId = setTimeout(() => {
      lastTimeoutId = null;
      fn(...args);
    }, timeoutMs);
  };
}

/** @hidden */
export function isBlockDevelopmentRestrictionEnabled() {
  return getAirtableInterface().sdkInitData.baseData.isBlockDevelopmentRestrictionEnabled;
}

/**
 * Added for use in Gantt View, to enable i18n.
 *
 * @hidden
 * */
export function getLocaleAndDefaultLocale() {
  const sdkInitData = getAirtableInterface().sdkInitData;
  return {
    locale: 'locale' in sdkInitData ? sdkInitData.locale : undefined,
    defaultLocale: 'defaultLocale' in sdkInitData ? sdkInitData.defaultLocale : undefined
  };
}

/** @hidden */
export { useMapApiToken as useMapApiToken_unstable } from '../interface/ui/use_map_api_token';
