// @airtable/blocks/dist/esm/shared/sdk_core.js
import GlobalConfig from './global_config';

/**
 * @hidden
 * @example
 * ```js
 * import {runInfo} from '@airtable/blocks/base';
 * if (runInfo.isFirstRun) {
 *     // The current user just installed this block.
 *     // Take the opportunity to show any onboarding and set
 *     // sensible defaults if the user has permission.
 *     // For example, if the block relies on a table, it would
 *     // make sense to set that to cursor.activeTableId
 * }
 * ```
 */

/** @hidden */
export class BlockSdkCore {
  /**
   * This value is used by the blocks-testing library to verify
   * compatibility.
   *
   * @hidden
   */
  // @ts-ignore
  static VERSION = "0.0.0-experimental-801a212b8-20260206";

  /** Storage for this block installation's configuration. */

  /** Contains information about the current session. */

  /** Represents the current Airtable {@link Base}. */

  /**
   * Returns the ID for the current block installation.
   *
   * @example
   * ```js
   * import {installationId} from '@airtable/blocks/base';
   * console.log(installationId);
   * // => 'blifDutUr92OKwnUn'
   * ```
   */

  /** @hidden */

  /** @internal */

  /** @internal */

  /** @hidden */
  constructor(airtableInterface) {
    this.__airtableInterface = airtableInterface;

    // @ts-ignore
    airtableInterface.assertAllowedSdkPackageVersion("@airtable/blocks", BlockSdkCore.VERSION);
    const sdkInitData = airtableInterface.sdkInitData;
    this.globalConfig = new GlobalConfig(sdkInitData.initialKvValuesByKey, this);
    this.installationId = sdkInitData.blockInstallationId;
    this.runInfo = Object.freeze({
      isFirstRun: sdkInitData.isFirstRun,
      isDevelopmentMode: sdkInitData.isDevelopmentMode,
      intentData: sdkInitData.intentData
    });
    this.session = this._constructSession();
    this.base = this._constructBase();
    this.__mutations = this._constructMutations();

    // Bind the public methods on this class so users can import
    // just the method, e.g.
    // import {reload} from '@airtable/blocks/base';
    this.reload = this.reload.bind(this);
  }

  /** @internal */

  /** @internal */

  /** @internal */

  /**
   * Call this function to reload your block.
   *
   * @example
   * ```js
   * import React from 'react';
   * import {reload} from '@airtable/blocks/base';
   * import {Button, initializeBlock} from '@airtable/blocks/base/ui';
   * function MyBlock() {
   *     return <Button onClick={() => reload()}>Reload</Button>;
   * }
   * initializeBlock(() => <MyBlock />);
   * ```
   */
  reload() {
    this.__airtableInterface.reloadFrame();
  }

  /**
   * @internal
   */
  get __appInterface() {
    return this.base._baseData.appInterface;
  }
}
