// @airtable/blocks/dist/esm/interface/ui/block_wrapper.js
/** @hidden */ /** */
import * as React from 'react';
import Loader from '../../shared/ui/loader';
import { SdkContext } from '../../shared/ui/sdk_context';
import { getCssContentToAddToHead, SPIN_SCALE_ANIMATION_NAME } from './global_css_helpers';
const suspenseFallbackStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
};
const animateSpinnerStyle = {
  animationIterationCount: 'infinite',
  animationName: SPIN_SCALE_ANIMATION_NAME,
  animationDuration: '1800ms',
  animationTimingFunction: 'cubic-bezier(0.785, 0.135, 0.15, 0.86)'
};
export const BlockWrapper = _ref => {
  let {
    sdk,
    children
  } = _ref;
  React.useLayoutEffect(() => {
    // Create a style element and inject the global CSS into document head
    const styleElement = document.createElement('style');
    styleElement.textContent = getCssContentToAddToHead();
    document.head.appendChild(styleElement);

    // Cleanup: remove the style element when component unmounts
    return () => {
      if (document.head.contains(styleElement)) {
        document.head.removeChild(styleElement);
      }
    };
  }, []);
  return /*#__PURE__*/React.createElement(SdkContext.Provider, {
    value: sdk
  }, /*#__PURE__*/React.createElement(React.Suspense, {
    fallback: /*#__PURE__*/React.createElement("div", {
      style: suspenseFallbackStyle
    }, /*#__PURE__*/React.createElement(Loader, {
      style: animateSpinnerStyle
    }))
  }, children));
};
