// @airtable/blocks/dist/esm/interface/ui/selectable_wrapper.js
/** @module @airtable/blocks/interface/ui: SelectableWrapper */ /** */
import React, { useContext, useEffect, useMemo, useRef, useState } from 'react';
import ReactDOM from 'react-dom';
import { useSdk } from '../../shared/ui/sdk_context';
import { useRunInfo } from './use_run_info';

/**
 * Context to track the nesting depth of SelectableWrapper components.
 * Used to ensure child overlays render above parent overlays via z-index.
 *
 * @internal
 */
const NestingDepthContext = /*#__PURE__*/React.createContext(0);

/** @hidden */

/**
 * A wrapper component that makes its child selectable with a click overlay.
 * When clicked, the child component is marked as selected. If already selected,
 * the overlay becomes click-through. Selected elements display a blue border.
 *
 * @component
 * @hidden
 */
export function SelectableWrapper(_ref) {
  let {
    children,
    sourceLocation,
    name
  } = _ref;
  const runInfo = useRunInfo();
  if (runInfo.isPageElementInEditMode) {
    return /*#__PURE__*/React.createElement(SelectableWrapperInEditMode, {
      sourceLocation: sourceLocation,
      name: name
    }, children);
  } else {
    return /*#__PURE__*/React.createElement(SelectableWrapperInViewMode, {
      name: name
    }, children);
  }
}

/** @internal */
function SelectableWrapperInViewMode(_ref2) {
  let {
    children
  } = _ref2;
  return children;
}

/** @internal */
function SelectableWrapperInEditMode(_ref3) {
  let {
    children,
    sourceLocation,
    name
  } = _ref3;
  const id = useUniqueId();
  const [ref, setRef] = useState(null);
  const sdk = useSdk();

  // Get the current nesting depth from context
  const nestingDepth = useContext(NestingDepthContext);

  // Check if this wrapper is currently selected
  const selectionData = useSelectedSubElementId();
  const isSelected = selectionData.selectedSubElementId === id;

  // Handle overlay click
  const handleOverlayClick = async event => {
    event.preventDefault();
    event.stopPropagation();

    // Only handle selection if not already selected
    if (!isSelected) {
      await sdk.setSelectedSubElementAsync({
        subElementId: id,
        sourceLocation,
        name
      });
    }
  };
  useEffect(() => {
    if (!isSelected) {
      return () => {};
    }

    // Add a global click listener to deselect this SelectableWrapper if there
    // is a click outside it
    const globalClickListener = async event => {
      // Check if the click is outside this SelectableWrapper div
      const isClickOutside = event.target instanceof Node && ref && !ref.contains(event.target);

      // Check if the click is on another SelectableWrapper's overlay
      const isClickOnOtherWrapper = event.target instanceof HTMLElement && event.target.dataset.selectableWrapperOverlay === 'true';
      if (isClickOutside && !isClickOnOtherWrapper) {
        // In this case, notify liveapp to deselect this sub element. The deselection
        // originated from the SDK.
        await sdk.setSelectedSubElementAsync(null);
      }
    };
    const globalClickListenerOptions = {
      // Use capture phase to ensure the click is handled even if stopPropagation is called.
      // This is necessary because the click event is triggered by a click on some other overlay,
      // the overlay calls stopPropagation.
      capture: true
    };
    document.addEventListener('click', globalClickListener, globalClickListenerOptions);
    return () => {
      document.removeEventListener('click', globalClickListener, globalClickListenerOptions);
    };
  }, [sdk, isSelected, ref]);

  // Track the latest isSelected value in a ref for use in unmount cleanup
  const isSelectedRef = useRef(isSelected);
  // Keep the ref in sync with isSelected
  useEffect(() => {
    isSelectedRef.current = isSelected;
  }, [isSelected]);
  // If the component unmounts while selected, deselect it. We use a ref to track
  // isSelected so that changes to isSelected don't cause this effect to re-run.
  // This effect should only run its cleanup when the component actually unmounts.
  useEffect(() => {
    return () => {
      if (isSelectedRef.current) {
        sdk.setSelectedSubElementAsync(null);
      }
    };
  }, [sdk]);
  return /*#__PURE__*/React.createElement(NestingDepthContext.Provider, {
    value: nestingDepth + 1
  }, /*#__PURE__*/React.createElement("div", {
    ref: setRef,
    style: {
      display: 'contents'
    }
  }, children, /*#__PURE__*/React.createElement(Overlay, {
    wrapperDiv: ref,
    isSelected: isSelected,
    ariaLabel: `Select ${name || 'element'}`,
    onClick: handleOverlayClick,
    nestingDepth: nestingDepth
  })));
}

/**
 * Traverses ancestors once to gather both z-index and scrollable containers.
 * Returns the effective z-index and an array of scrollable ancestor elements.
 *
 * @internal
 */
function getAncestorInfo(element) {
  const scrollableAncestors = [];
  let effectiveZIndex = 0;
  let current = element.parentElement;
  while (current && current !== document.body) {
    const style = window.getComputedStyle(current);

    // Check for effective z-index (only if we haven't found one yet)
    if (effectiveZIndex === 0) {
      const zIndex = parseInt(style.zIndex, 10);
      if (!isNaN(zIndex) && style.position !== 'static') {
        effectiveZIndex = zIndex;
      }
    }

    // Check if this ancestor is scrollable
    const isScrollable = [style.overflow, style.overflowX, style.overflowY].some(overflow => overflow === 'auto' || overflow === 'scroll');
    if (isScrollable) {
      scrollableAncestors.push(current);
    }
    current = current.parentElement;
  }
  return {
    effectiveZIndex,
    scrollableAncestors
  };
}

/** @internal */
function Overlay(_ref4) {
  let {
    wrapperDiv,
    isSelected,
    ariaLabel,
    onClick,
    nestingDepth
  } = _ref4;
  const [boundingClientRect, setBoundingClientRect] = useState(null);

  // Limitation: We currently only compute this once (for a given identity
  // of wrapperDiv), which means that
  // - if wrapperDiv or an ancestor is reparented, or
  // - if the z-index or scrollability of an ancestor changes
  // then these values will be stale.
  // In practice, reparenting DOM elements is rare in React apps.
  // TBD how common z-index or scrollability changes are in practice - it's
  // something we should keep an eye on.
  const {
    baseZIndex,
    scrollableAncestors
  } = useMemo(() => {
    if (!wrapperDiv) {
      return {
        baseZIndex: 0,
        scrollableAncestors: []
      };
    }
    const {
      effectiveZIndex,
      scrollableAncestors
    } = getAncestorInfo(wrapperDiv);
    return {
      baseZIndex: effectiveZIndex,
      scrollableAncestors
    };
  }, [wrapperDiv]);
  useEffect(() => {
    if (!wrapperDiv) {
      return () => {};
    }

    // Function to compute the aggregate bounding client rect
    const computeBoundingRect = () => {
      if (!wrapperDiv) {
        return null;
      }
      let aggregateBoundingClientRect = null;
      for (const child of wrapperDiv.children) {
        const boundingClientRect = child.getBoundingClientRect();
        const scrollLeft = window.pageXOffset || document.documentElement.scrollLeft;
        const scrollTop = window.pageYOffset || document.documentElement.scrollTop;
        if (!aggregateBoundingClientRect) {
          aggregateBoundingClientRect = {
            left: boundingClientRect.left + scrollLeft,
            top: boundingClientRect.top + scrollTop,
            right: boundingClientRect.right + scrollLeft,
            bottom: boundingClientRect.bottom + scrollTop
          };
          continue;
        }
        if (boundingClientRect.left + scrollLeft < aggregateBoundingClientRect.left) {
          aggregateBoundingClientRect.left = boundingClientRect.left + scrollLeft;
        }
        if (boundingClientRect.top + scrollTop < aggregateBoundingClientRect.top) {
          aggregateBoundingClientRect.top = boundingClientRect.top + scrollTop;
        }
        if (boundingClientRect.right + scrollLeft > aggregateBoundingClientRect.right) {
          aggregateBoundingClientRect.right = boundingClientRect.right + scrollLeft;
        }
        if (boundingClientRect.bottom + scrollTop > aggregateBoundingClientRect.bottom) {
          aggregateBoundingClientRect.bottom = boundingClientRect.bottom + scrollTop;
        }
      }
      return aggregateBoundingClientRect;
    };

    // Initial computation
    const rect = computeBoundingRect();
    setBoundingClientRect(rect);

    // Debounce helper
    let resizeTimeoutId = null;
    const debouncedUpdateRect = () => {
      // Immediately reset the bounding client rect to null when the window is resized
      // to avoid the overlay being in the wrong position as the resize is happening.
      // Then debounce the actual measurement with setTimeout for performance.
      setBoundingClientRect(null);
      if (resizeTimeoutId !== null) {
        clearTimeout(resizeTimeoutId);
      }
      resizeTimeoutId = setTimeout(() => {
        const updatedRect = computeBoundingRect();
        setBoundingClientRect(updatedRect);
        resizeTimeoutId = null;
      }, 16); // ~60fps
    };

    // Set up ResizeObserver to watch for size changes (debounced)
    const resizeObserver = new ResizeObserver(debouncedUpdateRect);

    // Observe resize on the wrapper element and all its children
    resizeObserver.observe(wrapperDiv);
    for (const child of wrapperDiv.children) {
      resizeObserver.observe(child);
    }

    // Set up MutationObserver to watch for changes to children
    const mutationObserver = new MutationObserver(() => {
      const updatedRect = computeBoundingRect();
      setBoundingClientRect(updatedRect);

      // Re-observe all children with the ResizeObserver in case new children were added
      // This ensures dynamically added children are also observed for resize events
      resizeObserver.disconnect();
      resizeObserver.observe(wrapperDiv);
      for (const child of wrapperDiv.children) {
        resizeObserver.observe(child);
      }
    });

    // Observe changes to child list and subtree (e.g., attributes, content)
    mutationObserver.observe(wrapperDiv, {
      childList: true,
      subtree: true,
      attributes: true,
      characterData: true
    });

    // Listen for window resize events to catch viewport changes (debounced)
    window.addEventListener('resize', debouncedUpdateRect);

    // Listen for scroll events on scrollable ancestor containers
    // Note: We don't need to listen to window scroll because the overlay and wrapped
    // element are both positioned relative to the document, so they scroll together naturally.
    // We only need to listen to scrollable containers because the overlay is portaled
    // outside them to document.body, so it won't scroll with them automatically.
    // Use passive: true for better scroll performance
    scrollableAncestors.forEach(scrollable => {
      scrollable.addEventListener('scroll', debouncedUpdateRect, {
        passive: true
      });
    });
    return () => {
      if (resizeTimeoutId !== null) {
        clearTimeout(resizeTimeoutId);
      }
      mutationObserver.disconnect();
      resizeObserver.disconnect();
      window.removeEventListener('resize', debouncedUpdateRect);
      scrollableAncestors.forEach(scrollable => {
        scrollable.removeEventListener('scroll', debouncedUpdateRect);
      });
    };
  }, [wrapperDiv, scrollableAncestors]);
  const [isHovered, setIsHovered] = useState(false);
  const handleMouseEnter = () => {
    setIsHovered(true);
  };
  const handleMouseLeave = () => {
    setIsHovered(false);
  };
  if (!boundingClientRect) {
    return null;
  }
  return /*#__PURE__*/ReactDOM.createPortal( /*#__PURE__*/React.createElement("div", {
    onClick: onClick,
    role: "button"
    // Prevent keyboard focus. This is consistent with the behavior of native interface elements
    // in interface designer. See tk tk
    ,
    tabIndex: -1,
    "data-selectable-wrapper-overlay": "true",
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    "aria-label": ariaLabel,
    style: {
      position: 'absolute',
      left: boundingClientRect.left,
      top: boundingClientRect.top,
      height: boundingClientRect.bottom - boundingClientRect.top,
      width: boundingClientRect.right - boundingClientRect.left,
      backgroundColor: isSelected ? 'transparent' : isHovered ? 'rgba(22, 110, 225, 0.05)' : 'transparent',
      cursor: isSelected ? 'default' : 'pointer',
      pointerEvents: isSelected ? 'none' : 'auto',
      outline: isSelected ? '2px solid #3b82f6' : undefined,
      outlineOffset: isSelected ? '-2px' : undefined,
      // borderRadius: isSelected ? '4px' : undefined,
      // Use base z-index from ancestors + 1 to be above the wrapper + nestingDepth for nested overlays
      zIndex: baseZIndex + 1 + nestingDepth
    }
  }), document.body);
}

/** @internal */
function useUniqueId() {
  return useMemo(() => Math.random().toString(36).substring(2, 15), []);
}

/** @internal */
function useSelectedSubElementId() {
  const sdk = useSdk();
  const [selectedSubElementId, setSelectedSubElementId] = useState(null);
  useEffect(() => {
    let isCancelled = false;
    // Subscribe to selection data updates
    const handleSelectionUpdate = data => {
      setSelectedSubElementId(data.selectedSubElementId);
    };

    // Fetch initial selection data and subscribe to updates

    sdk.fetchAndSubscribeToSelectionDataAsync(handleSelectionUpdate).then(initialData => {
      if (!isCancelled) {
        setSelectedSubElementId(initialData.selectedSubElementId ?? null);
      }
    });

    // Cleanup: unsubscribe when component unmounts
    return () => {
      isCancelled = true;
      sdk.unsubscribeFromSelectionData();
    };
  }, [sdk]);
  return {
    selectedSubElementId
  };
}
