// @airtable/blocks/dist/esm/interface/ui/use_map_api_token.js
import { useCallback, useEffect, useState } from 'react';
import { useSdk } from '../../shared/ui/sdk_context';

/** @hidden */

/**
 * @hidden
 * Experimental implementation of using an Airtable-provided token
 */
export function useMapApiToken() {
  const [fetchState, setFetchState] = useState({
    status: 'initial'
  });
  const sdk = useSdk();
  const fetchApiToken = useCallback(async () => {
    try {
      const getTokenResponse = await sdk.unstable_getMapApiTokenAsync();
      if (getTokenResponse.success) {
        setFetchState({
          status: 'completed',
          token: getTokenResponse.apiKey,
          warningMessage: getTokenResponse.warningMessage
        });
      } else {
        setFetchState({
          status: 'error',
          errorMessage: getTokenResponse.userFriendlyErrorMessage,
          suggestedPrompt: getTokenResponse.suggestedOmniPrompt
        });
      }
    } catch (err) {
      setFetchState({
        status: 'error',
        errorMessage: `An unexpected error occurred while loading the map API token:\n${err}`
      });
    }
  }, [sdk]);
  useEffect(() => {
    if (fetchState.status === 'initial') {
      setFetchState({
        status: 'loading'
      });
      fetchApiToken();
    }
  }, [fetchApiToken, fetchState.status]);
  return fetchState;
}
