// @airtable/blocks/dist/esm/interface/sdk.js
import { BlockSdkCore } from '../shared/sdk_core';
import { invariant } from '../shared/error_utils';
import { Session } from './models/session';
import { Mutations } from './models/mutations';
import { Base } from './models/base';
/** @hidden */
export class InterfaceBlockSdk extends BlockSdkCore {
  constructor(airtableInterface) {
    super(airtableInterface);
    this._registerHandlers();
  }
  /** @internal */
  _constructSession() {
    return new Session(this);
  }
  /** @internal */
  _constructBase() {
    return new Base(this);
  }
  /** @internal */
  _constructMutations() {
    return new Mutations(this, this.session, this.base, changes => this.__applyModelChanges(changes), updates => this.__applyGlobalConfigUpdates(updates));
  }
  /** @internal */
  _registerHandlers() {
    // base
    this.__airtableInterface.subscribeToModelUpdates(_ref => {
      let {
        changes
      } = _ref;
      this.__applyModelChanges(changes);
    });

    // global config
    this.__airtableInterface.subscribeToGlobalConfigUpdates(_ref2 => {
      let {
        updates
      } = _ref2;
      this.__applyGlobalConfigUpdates(updates);
    });
  }
  /** @internal */
  __applyModelChanges(changes) {
    const changedBasePaths = this.base.__applyChangesWithoutTriggeringEvents(changes);
    const changedSessionKeys = this.session.__applyChangesWithoutTriggeringEvents(changes);
    this.base.__triggerOnChangeForChangedPaths(changedBasePaths);
    this.session.__triggerOnChangeForChangedKeys(changedSessionKeys);
  }
  /** @internal */
  __applyGlobalConfigUpdates(updates) {
    this.globalConfig.__setMultipleKvPaths(updates);
  }

  /**
   * @internal
   */
  get __appInterface() {
    return this.base._baseData.appInterface;
  }

  /** @hidden */
  getBlockRunContext() {
    return this.__airtableInterface.sdkInitData.runContext;
  }

  /**
   * @internal
   */
  setCustomPropertiesAsync(properties) {
    return this.__airtableInterface.setCustomPropertiesAsync(properties);
  }

  /**
   * @internal
   */
  setSelectedSubElementAsync(selectedSubElement) {
    return this.__airtableInterface.setSelectedSubElementAsync(selectedSubElement);
  }

  /**
   * @internal
   */
  fetchAndSubscribeToSelectionDataAsync(callback) {
    return this.__airtableInterface.fetchAndSubscribeToSelectionDataAsync(callback);
  }

  /**
   * @internal
   */
  unsubscribeFromSelectionData() {
    return this.__airtableInterface.unsubscribeFromSelectionData();
  }

  /**
   * @hidden
   */
  unstable_getMapApiTokenAsync() {
    invariant(this.__airtableInterface.getMapApiTokenAsync, 'map API token is not supported in this context');
    return this.__airtableInterface.getMapApiTokenAsync();
  }
}
