// @airtable/blocks/dist/esm/interface/types/airtable_interface.js
// BlockRunContextType, BlockInstallationPageBlockRunContext, BlockRunContext
// must be kept in sync with block_run_context.tsx in hyperbase repo
/** @hidden */
export let BlockRunContextType = /*#__PURE__*/function (BlockRunContextType) {
  BlockRunContextType["PAGE_ELEMENT_IN_QUERY_CONTAINER"] = "pageElementInQueryContainer";
  return BlockRunContextType;
}({});

/** @hidden */

/** @hidden */

/** @hidden */

/** @hidden */

/** @hidden */

// Keep in sync with enum in hyperbase repo:
// https://github.com/Hyperbase/hyperbase/blob/2c0e2ef2cbc6bce7ab33fd913ef3572a83c3892d/client_server_shared/pages/types/block_installation_page_elements/block_installation_page_element_custom_properties_types.tsx
/** @hidden */
export let BlockInstallationPageElementCustomPropertyTypeForAirtableInterface = /*#__PURE__*/function (BlockInstallationPageElementCustomPropertyTypeForAirtableInterface) {
  BlockInstallationPageElementCustomPropertyTypeForAirtableInterface["BOOLEAN"] = "boolean";
  BlockInstallationPageElementCustomPropertyTypeForAirtableInterface["STRING"] = "string";
  BlockInstallationPageElementCustomPropertyTypeForAirtableInterface["ENUM"] = "enum";
  BlockInstallationPageElementCustomPropertyTypeForAirtableInterface["FIELD_ID"] = "fieldId";
  BlockInstallationPageElementCustomPropertyTypeForAirtableInterface["TABLE_ID"] = "tableId";
  return BlockInstallationPageElementCustomPropertyTypeForAirtableInterface;
}({});

// Keep in sync with type in hyperbase repo:
// https://github.com/Hyperbase/hyperbase/blob/2c0e2ef2cbc6bce7ab33fd913ef3572a83c3892d/client_server_shared/pages/types/block_installation_page_elements/block_installation_page_element_custom_properties_types.tsx
/** @hidden */

/** @hidden */

/** @hidden */

// This type should be kept in sync with the hyperbase repo's fetch_map_api_key_types.tsx
/** @hidden */

/**
 * AirtableInterface is designed as the communication interface between the
 * Block SDK and Airtable.
 *
 * @hidden
 */
