// @airtable/blocks/dist/esm/testing/base/abstract_mock_airtable_interface.js
// istanbul ignore file
import EventEmitter from 'events';
import { cloneDeep } from '../../shared/private_utils';
import { spawnError } from '../../shared/error_utils';
/** @internal */
const aggregators = {
  aggregate(appInterface, summaryFunctionKey, cellValues, fieldData) {
    return null;
  },
  aggregateToString(appInterface, summaryFunctionKey, cellValues, fieldData) {
    return '';
  },
  getAggregatorConfig(summaryFunctionKey) {
    return {
      key: '',
      displayName: '',
      shortDisplayName: ''
    };
  },
  getAllAvailableAggregatorKeys() {
    return [];
  },
  getAvailableAggregatorKeysForField(fieldData) {
    return [];
  }
};

/** @internal */
const fieldTypeProvider = {
  isComputed(fieldData) {
    return false;
  },
  validateConfigForUpdate: () => {
    return {
      isValid: true
    };
  },
  getConfig: (appInterface, fieldData, fieldNamesById) => {
    return {
      type: fieldData.type,
      options: fieldData.typeOptions
    };
  },
  canBePrimary: () => {
    return true;
  },
  convertStringToCellValue(appInterface, string, fieldData) {
    return '';
  },
  convertCellValueToString(appInterface, cellValue, fieldData) {
    return '';
  },
  getCellRendererData(appInterface, cellValue, fieldData, shouldWrap) {
    return {
      cellValueHtml: `<pre>${JSON.stringify(cellValue)}</pre>`,
      attributes: {}
    };
  },
  validateCellValueForUpdate(appInterface, newCellValue, currentCellValue, fieldData) {
    return {
      isValid: true
    };
  },
  getUiConfig(appInterface, fieldData) {
    return {
      iconName: '',
      desiredCellWidthForRecordCard: 0,
      minimumCellWidthForRecordCard: 0
    };
  }
};

/** @internal */
const urlConstructor = {
  getTableUrl(tableId) {
    return `https://airtable.test/${tableId}`;
  },
  getViewUrl(viewId, tableId) {
    return `https://airtable.test/${tableId}/${viewId}`;
  },
  getRecordUrl(recordId, tableId) {
    return `https://airtable.test/${tableId}/${recordId}`;
  },
  getAttachmentClientUrl(appInterface, attachmentId, attachmentUrl) {
    return attachmentUrl;
  }
};

/** @internal */
const globalConfigHelpers = {
  validatePath(path, store) {
    return {
      isValid: true
    };
  },
  validateAndApplyUpdates(updates, store) {
    throw spawnError('validateAndApplyUpdates unimplemented');
  }
};

/** @internal */
const idGenerator = {
  generateRecordId: () => 'recGeneratedMockId',
  generateFieldId: () => 'fldGeneratedMockId',
  generateTableId: () => 'tblGeneratedMockId'
};

/**
 * An abstract base class with a common interface exposed to both Blocks SDK's
 * internal automated test suite and the blocks-testing public repo.
 *
 * @hidden
 */
export class AbstractMockAirtableInterface extends EventEmitter {
  constructor(initData) {
    super();
    this._initData = cloneDeep(initData);
    this.reset();
  }

  /**
   * Revert the mock interface to its initial state. This includes:
   *
   * - removing all event listeners
   * - restoring the database schema
   */
  reset() {
    this.removeAllListeners();
    this.sdkInitData = cloneDeep(this._initData);
  }

  // The following properties are defined using accessors in order to
  // facilitate overriding in subclasses.
  get aggregators() {
    return aggregators;
  }
  get fieldTypeProvider() {
    return fieldTypeProvider;
  }
  get urlConstructor() {
    return urlConstructor;
  }
  get globalConfigHelpers() {
    return globalConfigHelpers;
  }
  get idGenerator() {
    return idGenerator;
  }
  assertAllowedSdkPackageVersion() {}
  applyMutationAsync(mutation, opts) {
    return Promise.resolve();
  }
  checkPermissionsForMutation(mutation) {
    return {
      hasPermission: true
    };
  }
  createVisList(appInterface, recordDatas, fieldDatas, sorts) {
    return {
      removeRecordIds(recordIds) {},
      addRecordData(recordData) {},
      getOrderedRecordIds() {
        return recordDatas.map(_ref => {
          let {
            id
          } = _ref;
          return id;
        });
      }
    };
  }
  subscribeToModelUpdates(fn) {
    this.on('modelupdates', fn);
  }
  setActiveViewOrTable(tableId, viewId) {}

  // TODO(jugglinmike): Implement
  subscribeToGlobalConfigUpdates() {}
  subscribeToSettingsButtonClick() {}
  subscribeToEnterFullScreen() {}
  subscribeToExitFullScreen() {}
  subscribeToFocus() {}
  triggerModelUpdates(changes) {
    this.emit('modelupdates', {
      changes
    });
  }

  // TODO(jugglinmike): Implement
  triggerGlobalConfigUpdates() {}
  triggerSettingsButtonClick() {}
  triggerEnterFullScreen() {}
  triggerExitFullScreen() {}
  triggerFocus() {}
  unsubscribeFromCursorData() {}
  unsubscribeFromTableData() {}
  unsubscribeFromCellValuesInFields() {}
  unsubscribeFromViewData() {}
}
